sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java

```diff
@@ -0,0 +1,1071 @@
+     */
+    public <K, DataT, InputT, OutputT> ComposedKeyedCombineFnWithContext<DataT, K> with(
+        SimpleFunction<DataT, InputT> extractInputFn,
+        CombineFnWithContext<InputT, ?, OutputT> combineFnWithContext,
+        TupleTag<OutputT> outputTag) {
+      return with(extractInputFn, combineFnWithContext.<K>asKeyedFn(), outputTag);
+    }
+  }
+
+  /**
+   * A builder class to construct a composed {@link GlobalCombineFn}.
+   */
+  public static class ComposeCombineFnBuilder {
+    /**
+     * Returns a {@link ComposedCombineFn} that can take additional
+     * {@link GlobalCombineFn GlobalCombineFns} and apply them as a single combine function.
+     *
+     * <p>The {@link ComposedCombineFn} extracts inputs from {@code DataT} with
+     * the {@code extractInputFn} and combines them with the {@code combineFn},
+     * and then it outputs each combined value with a {@link TupleTag} to a
+     * {@link CoCombineResult}.
+     */
+    public <DataT, InputT, OutputT> ComposedCombineFn<DataT> with(
+        SimpleFunction<DataT, InputT> extractInputFn,
+        CombineFn<InputT, ?, OutputT> combineFn,
+        TupleTag<OutputT> outputTag) {
+      return new ComposedCombineFn<DataT>()
+          .with(extractInputFn, combineFn, outputTag);
+    }
+
+    /**
+     * Returns a {@link ComposedCombineFnWithContext} that can take additional
+     * {@link GlobalCombineFn GlobalCombineFns} and apply them as a single combine function.
+     *
+     * <p>The {@link ComposedCombineFnWithContext} extracts inputs from {@code DataT} with
+     * the {@code extractInputFn} and combines them with the {@code combineFnWithContext},
+     * and then it outputs each combined value with a {@link TupleTag} to a
+     * {@link CoCombineResult}.
+     */
+    public <DataT, InputT, OutputT> ComposedCombineFnWithContext<DataT> with(
+        SimpleFunction<DataT, InputT> extractInputFn,
+        CombineFnWithContext<InputT, ?, OutputT> combineFnWithContext,
+        TupleTag<OutputT> outputTag) {
+      return new ComposedCombineFnWithContext<DataT>()
+          .with(extractInputFn, combineFnWithContext, outputTag);
+    }
+  }
+
+  /////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * A tuple of {@link Object Objects} produced for a given key from the composed combine functions.
+   */
+  public static class CoCombineResult implements Serializable {
+    private final Map<TupleTag<?>, Object> valuesMap;
+    private final Set<TupleTag<?>> tupleTags;
+
+    /**
+     * The constructor of {@link CoCombineResult}.
+     *
+     * @throws NullPointerException if any key or value in {@code mavaluesMapp} is null
+     */
+    CoCombineResult(
+        Map<TupleTag<?>, Object> valuesMap, Iterable<TupleTag<?>> tupleTags) {
+      this.valuesMap = ImmutableMap.copyOf(valuesMap);
+      this.tupleTags = ImmutableSet.copyOf(tupleTags);
```

Comment:
I think ImmutableSet has the same problem, so I'm guessing that the problem is with the value? It may be clearer to have allocate a static "NULL_VALUE" Object, and use that as a sentinel so you can put a non-nvull value in, and then use the single valuesMap everywhere.

If not, I think making it really explicit (`// Store tupleTags separately since tupleTags with null values are omitted`) would be good.